Feature/tube

hanhanW opened the pull request:
1. Implement tube and sock

david942j commented on test/dynelf_test.rb:
```diff
@@ -37,8 +37,8 @@ def test_lookup
-        h.each do |sym, off|
-          assert_includes(off, d.lookup(sym) - d.libbase)
+        %w(system open read write execve printf puts sprintf mmap mprotect).each do |sym|
+          assert_includes(h[sym], d.lookup(sym) - d.libbase)
```
This change should not present in this branch

hanhanW replied:
You're right, fixed.

david942j replied:
seems not fixed?

david942j commented on test/timer_test.rb:
```diff
@@ -0,0 +1,21 @@
+
+  def test_countdown
+    t = Timer.new
+    assert_nil(t.started?)
```
Use refute might be clearer?

hanhanW commented:
Review status: 0 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[test/timer_test.rb, line 12 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fxveve7T0CHwbG-r2-12:-KrLfBpCWHZPvgI0KpnI:bqog7zb) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/9c1fa7e4ef6235daeae07040d777bbadd6ba24bc/test/timer_test.rb#L12)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

Use refute might be clearer?
</blockquote></details>

Agree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+      # @raise [ArgumentError]
+      #   If the block is not given.
+      def recvpred(timeout: nil)
+        raise ArgumentError, 'recvpred with no pred' unless block_given?
```
let the error message clearer
ex. `Need a block for recvpred`

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+      #   waiting.
+      def recvn(num_bytes, timeout: nil)
+        @timer.countdown(timeout) do
+          # TODO(Darkpi): Select!
```
this todo seems important (?

david942j commented on lib/pwnlib/tubes/sock.rb:
```diff
@@ -0,0 +1,66 @@
+      #
+      # @param [String] host
+      #   The host to connect.
+      #
```
No new line here

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+      #       since the latter can't be done greedly. Still, it would be bad to call this
+      #       for case with ambiguity.
+      #
+      # @param [Regexp] regex
```
wrong doc

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+      #   A regex to match.
+      # @param [Float] timeout
+      #   Any positive float, indicates timeouts in seconds.
+      #
```
doc for `drop`

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+      end
+
+      # Receive a single line from the tube.
+      # A "line" is any sequence of bytes terminated by the byte sequence set in +newline+, which
```
set in `+context.newline+`

david942j commented on lib/pwnlib/tubes/tube.rb:
```diff
@@ -0,0 +1,252 @@
+module Pwnlib
+  module Tubes
+    # Things common to all tubes (sockets, tty, ...)
+    class Tube
```
add TODO for using hexdump in send/recv

david942j commented on test/timer_test.rb:
```diff
@@ -0,0 +1,21 @@
+  def test_countdown
+    t = Timer.new
+    refute(t.started?)
+    refute(t.active?)
```
add test case for when will these two methods return true.

hanhanW commented:
Review status: 0 of 16 files reviewed at latest revision, 23 unresolved discussions.

---

*[.pullapprove.yml, line 14 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrLcyQ8r_Dpqdh9Sfxl-r4-14:-KrPXABFPiTnqZ0LN7v8:b-pplrac) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/.pullapprove.yml#L14)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

why does this change present here?
</blockquote></details>

Github didnt update `Files Changed` after rebase.
It's better to use `Reviewable`.

---

*[lib/pwnlib/tubes/tube.rb, line 64 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-64:-KrPYQb1aJjwesU0-Wr0:b-lwafy2) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L64)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

what does `pred` stand for?
</blockquote></details>

wiki: In mathematical logic, a predicate is commonly understood to be a Boolean-valued function.

---

*[test/dynelf_test.rb, line 41 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fxveve7T0CHwbD-r2-41:-KrPYafuj3nq9OgXz0mn:b-6fypjy) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/9c1fa7e4ef6235daeae07040d777bbadd6ba24bc/test/dynelf_test.rb#L41)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

seems not fixed?
</blockquote></details>

Use `Reviewable`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

hanhanW commented:
Review status: 0 of 16 files reviewed at latest revision, 23 unresolved discussions.

---

*[lib/pwnlib/tubes/buffer.rb, line 7 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGl-r4-7:-KrfkUyFPeWK8dxNLHsZ:b-uyp674) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/buffer.rb#L7)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

why `an` instead of `a`
</blockquote></details>

Fixed. Same problem in sock.rb and tube.rb are fixed, too.

---

*[lib/pwnlib/tubes/buffer.rb, line 23 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGl-r4-23:-KrfkpY4BeGWEFZcHpaf:b-886gsx) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/buffer.rb#L23)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

too -> either
</blockquote></details>

Fixed

---

*[lib/pwnlib/tubes/buffer.rb, line 65 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGl-r4-65:-KrflEuIHpgj-R64HqgF:b-886gsx) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/buffer.rb#L65)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`@return [String]`
</blockquote></details>

Fixed

---

*[lib/pwnlib/tubes/sock.rb, line 15 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-15:-KrflP25x9x7d4Lws_pT:b-886gsx) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L15)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

No new line here
</blockquote></details>

Fixed

---

*[lib/pwnlib/tubes/sock.rb, line 30 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-30:-Krfm_W-yDE3Yby65A_H:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L30)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

define method `close` as our previous discussion
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/sock.rb, line 53 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-53:-KrfmgSwg7VWKgkv8DiN:b-886gsx) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L53)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

seems this loop is redundant?
</blockquote></details>

Fixed

---

*[lib/pwnlib/tubes/tube.rb, line 10 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-10:-KrfnJ0Xbz4Ps3LuwODD:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L10)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

add TODO for using hexdump in send/recv
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/tube.rb, line 16 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-16:-KrfrDnSXoU6naiB12Iy:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L16)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

Doc the meaning when `timeout` equals to `nil`
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/tube.rb, line 27 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-27:-KrfrG2UEpkMT4I_AD0J:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L27)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

define a macro for `@param timeout`?
It appears many times with same description
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/tube.rb, line 31 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-31:-KrfrtdCGIVE-eDNaD9I:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L31)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`contains`
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/tube.rb, line 47 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-47:-KrfvC7GlGIGB5i3MBjw:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L47)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

what's `+pred(bytes)+ `
</blockquote></details>

done

---

*[lib/pwnlib/tubes/tube.rb, line 53 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-53:-KrfvEJTgBVi12Q3sprI:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L53)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

why wrap 'pred' with `+`
</blockquote></details>

done

---

*[lib/pwnlib/tubes/tube.rb, line 56 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-56:-KrfvH3WJiiAAE6-zIt9:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L56)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

same here
</blockquote></details>

done

---

*[lib/pwnlib/tubes/tube.rb, line 65 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-65:-KrfvTept3dXUv4EtUVX:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L65)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

let the error message clearer
ex. `Need a block for recvpred`
</blockquote></details>

done

---

*[lib/pwnlib/tubes/tube.rb, line 103 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-103:-Krg3HHZrBl-wU3eMtkU:b-l3tdk7) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L103)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

this todo seems important (?
</blockquote></details>

I traced the code and found it's blocking IO.
It sould be blocked in `recv_raw`.

---

*[lib/pwnlib/tubes/tube.rb, line 113 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-113:-KrfxJbzB9lpuI7D8OQS:b-7ql1td) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L113)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

wrong doc
</blockquote></details>

fixed

---

*[lib/pwnlib/tubes/tube.rb, line 117 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-117:-KrfxKYEr4MjfhA1lMXw:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L117)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

doc for `drop`
</blockquote></details>

done

---

*[lib/pwnlib/tubes/tube.rb, line 163 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGn-r4-163:-KrfxMICnZdNmFaXZbtM:b-7ql1td) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/tube.rb#L163)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

set in `+context.newline+`
</blockquote></details>

fixed

---

*[test/timer_test.rb, line 13 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fxveve7T0CHwbG-r4-13:-Krg1mG9UhXW9FUSCHgN:bb7f2mf) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/test/timer_test.rb#L13)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

add test case for when will these two methods return true.
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
Reviewed 3 of 7 files at r1, 3 of 10 files at r2, 1 of 3 files at r4, 3 of 7 files at r5.
Review status: 10 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 30 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-30:-KrkDmZLF3svqt7m4JF3:b-3i5kpg) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L30)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

Done
</blockquote></details>

also with `close(:write) / close(:send) / close(:read) / close(:recv)` to close one side only

---

*[.pullapprove.yml, line 14 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrLcyQ8r_Dpqdh9Sfxl-r4-14:-KrkHHW7_d49ia5n-1gk:b-gl29u4) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/.pullapprove.yml#L14)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

Github didnt update `Files Changed` after rebase.
It's better to use `Reviewable`.
</blockquote></details>

But I also see this change in Reviewable..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

hanhanW commented:
Review status: 10 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 30 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-30:-KrkjmEefQU5OAjK1t0r:b9629ir) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L30)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

also with `close(:write) / close(:send) / close(:read) / close(:recv)` to close one side only
</blockquote></details>

Pwntools implements in `shutdown`, I added TODO there.

In my opinion, it's more clear to implement in `close`. Maybe we should add TODO here?

---

*[.pullapprove.yml, line 14 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrLcyQ8r_Dpqdh9Sfxl-r4-14:-Krkl4gZsGxrbj7R6iyK:byqghkz) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/.pullapprove.yml#L14)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

But I also see this change in Reviewable..
</blockquote></details>

I dont know why, but it seems good here.

https://github.com/peter50216/pwntools-ruby/blame/feature/tube/.pullapprove.yml

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

hanhanW commented:
Review status: 6 of 14 files reviewed at latest revision, 4 unresolved discussions.

---

*[.pullapprove.yml, line 14 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrLcyQ8r_Dpqdh9Sfxl-r4-14:-Krq45_xdL7fWBJ3pYtE:b-ifdnsv) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/.pullapprove.yml#L14)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

I dont know why, but it seems good here.

https://github.com/peter50216/pwntools-ruby/blame/feature/tube/.pullapprove.yml

</blockquote></details>

Not work after merged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
Reviewed 1 of 2 files at r3, 3 of 5 files at r6.
Review status: 9 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 30 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-30:-KruVOzqgaRliZlKyAFb:b-3abc3p) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L30)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

Pwntools implements in `shutdown`, I added TODO there.

In my opinion, it's more clear to implement in `close`. Maybe we should add TODO here?
</blockquote></details>

No todo needed, but add `@diff` tag to document the difference between pwntools-python

---

*[lib/pwnlib/tubes/sock.rb, line 35 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVbSrpL9aRXWpPFFd:-KruVbSrpL9aRXWpPFFe:b-pb9bm7) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L35)):*
> ```Ruby
>       # @param [:both, :recv, :read, :send, :write]
>       #   Close the TCPSocket if +:both+ or no arguments passed.
>       #   Disallow further read in +sock+ if +:recv+ or +:read+ passed.
> ```

indent

---

*[lib/pwnlib/tubes/sock.rb, line 44 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVhlWCYQISOijr2gY:-KruVhlWCYQISOijr2gZ:b-kr71nx) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L44)):*
> ```Ruby
>           @sock.close
>         when :recv, :read
>           return if @closed[:recv]
> ```

`shutdown` will check this

---

*[lib/pwnlib/tubes/sock.rb, line 50 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVqVrz4pAF29IEZE4:-KruVqVrz4pAF29IEZE5:b-d5htnv) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L50)):*
> ```Ruby
>           shutdown(:send)
>         else
>           raise ArgumentError
> ```

add allowed options (:read, :recv, etc.) to error message

---

*[lib/pwnlib/tubes/tube.rb, line 124 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruWH8P_6ChZ3P3PXB-:-KruWH8P_6ChZ3P3PXB0:b-738573) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/tube.rb#L124)):*
> ```Ruby
>       #   Whether drop the ending.
>       # @!macro timeout_definition
>       #
> ```

missing `@return`

---

*[test/tubes/sock_test.rb, line 64 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruWVynn1BW8vSS0_yD:-KruWVynn1BW8vSS0_yE:b-6vx6os) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/test/tubes/sock_test.rb#L64)):*
> ```Ruby
>       assert_raises(EOFError) { s.recv }
>       assert_raises(EOFError) { s.recv }
>       3.times { s.close }
> ```

test of `s.close(:lalala)` ?

---

*[test/dynelf_test.rb, line 41 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fxveve7T0CHwbD-r2-41:-KruWew4-a9UWmwe34gV:b-ycq2in) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/9c1fa7e4ef6235daeae07040d777bbadd6ba24bc/test/dynelf_test.rb#L41)):*
<details><summary><i>Previously, hanhanW wrote…</i></summary><blockquote>

Use `Reviewable`.
</blockquote></details>

will here conflict with another pr (#35)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
Those issues have been resolved after I merged the master branch
What did you do before...

hanhanW commented:
Review status: 9 of 14 files reviewed at latest revision, 7 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 30 at r4](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fvMCeren-YYwGm-r4-30:-KrvFQkPmXVe1YZZQv-D:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/6a7818faea6190f35bd88bde19423a3e313e2591/lib/pwnlib/tubes/sock.rb#L30)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

No todo needed, but add `@diff` tag to document the difference between pwntools-python
</blockquote></details>

Done

---

*[lib/pwnlib/tubes/sock.rb, line 35 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVbSrpL9aRXWpPFFd:-KrvBMGgf1q7RJ8RJtbT:bk10jtu) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L35)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

indent
</blockquote></details>

Resolved with using bullet instead of indent.

---

*[lib/pwnlib/tubes/sock.rb, line 44 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVhlWCYQISOijr2gY:-KrvAT4k9fX02irObSpV:bt2vr1b) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L44)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

`shutdown` will check this
</blockquote></details>

Fixed.

---

*[lib/pwnlib/tubes/sock.rb, line 50 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruVqVrz4pAF29IEZE4:-KrvAtGpLUsUJcfkmIz6:b2nhexc) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/sock.rb#L50)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

add allowed options (:read, :recv, etc.) to error message
</blockquote></details>

Fix

---

*[lib/pwnlib/tubes/tube.rb, line 124 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruWH8P_6ChZ3P3PXB-:-KrvCGN5WKCRLWMLt6O8:b2nhexc) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/lib/pwnlib/tubes/tube.rb#L124)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

missing `@return`
</blockquote></details>

Fix

---

*[test/tubes/sock_test.rb, line 64 at r6](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KruWVynn1BW8vSS0_yD:-KrvBjmYek5QkW_XyDSi:bb6un1j) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/e6b11455f9e6a317fcc6fb13484779d85449d4b5/test/tubes/sock_test.rb#L64)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

test of `s.close(:lalala)` ?
</blockquote></details>

Done

---

*[test/dynelf_test.rb, line 41 at r2](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-KrHV5fxveve7T0CHwbD-r2-41:-KrvBkrb-a516GIRY1Pl:b-7zfwrl) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/9c1fa7e4ef6235daeae07040d777bbadd6ba24bc/test/dynelf_test.rb#L41)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

will here conflict with another pr (#35)?
</blockquote></details>

I'll merge master to pr(#35) after it's merged.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
Reviewed 1 of 10 files at r2, 2 of 7 files at r5, 1 of 3 files at r8.
Review status: 12 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 37 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1pmiEMHkMDewZbJio:-Ks1pmiEMHkMDewZbJip:b-v7fnbu) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/sock.rb#L37)):*
> ```Ruby
>       #   * Disallow further read in +sock+ if +:recv+ or +:read+ passed.
>       #   * Disallow further write in +sock+ if +:send+ or +:write+ passed.
>       # @diff In pwntools-python, method +shutdown(direction)+ is for closing socket one side,
> ```

one empty line

---

*[lib/pwnlib/tubes/tube.rb, line 63 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1rYHXYsdDkjPsIXCX:-Ks1rYHXYsdDkjPsIXCY:bsnq3zu) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/tube.rb#L63)):*
> ```Ruby
>       #   Whether the data satisfy the condition.
>       #
>       # @return [String]
> ```

move `@return` before `@yield`

---

*[lib/pwnlib/tubes/tube.rb, line 115 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1rOD70mFpY-LfxmXS:-Ks1rOD8CmyCxxKv-dNY:bbhjp6l) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/tube.rb#L115)):*
> ```Ruby
>       # +timeout+ seconds pass, all data is buffered and an empty string is returned.
>       #
>       # @diff We return the string that ends the earliest, rather then starts the earliest,
> ```

move `@diff` after `@return`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

hanhanW commented:
Review status: 12 of 14 files reviewed at latest revision, 3 unresolved discussions.

---

*[lib/pwnlib/tubes/sock.rb, line 37 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1pmiEMHkMDewZbJio:-Ks1xHA3GFL8Z9Tx2i5X:b2ni2nk) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/sock.rb#L37)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

one empty line
</blockquote></details>

fix

---

*[lib/pwnlib/tubes/tube.rb, line 63 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1rYHXYsdDkjPsIXCX:-Ks1zLCqNmaEZpcxII3Q:b2ni2nk) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/tube.rb#L63)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

move `@return` before `@yield`
</blockquote></details>

fix

---

*[lib/pwnlib/tubes/tube.rb, line 115 at r8](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-Ks1rOD70mFpY-LfxmXS:-Ks1xgdbDHbbtiqklbt1:b2ni2nk) ([raw file](https://github.com/peter50216/pwntools-ruby/blob/ba156b3bed8708d2fa4837503d093cef8e3a28c4/lib/pwnlib/tubes/tube.rb#L115)):*
<details><summary><i>Previously, david942j wrote…</i></summary><blockquote>

move `@diff` after `@return`
</blockquote></details>

fix

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29)*
<!-- Sent from Reviewable.io -->

david942j commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/peter50216/pwntools-ruby/29#-:-Ks3reTuN6thmm5Mslv0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

david942j commented:
But tube still not good enough to use without hex logger
